File: TestParser/TestUnion.fs
module TestParser.TestUnion

open NUnit.Framework
open CbStyles.Cbon.Parser
open CbStyles.Cbon.Parser.Parser
open CbStyles.Parser
open System.Collections.Generic


[<SetUp>]
let Setup () =
    ()

[<Test>]
let TestUnion1 () =
    let code = Reader.reader "(asd) 123"
    let r = union code
    printf "%s" (r.ToString())
    Assert.Pass()
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), CbAst.Union (AUnion("asd", CbAst.Num (ANum "123")))), r)

[<Test>]
let TestUnion2 () =
    let code = Reader.reader "('asd') (123) true"
    let r = union code
    printf "%s" (r.ToString())
    Assert.Pass()
    Assert.AreEqual(ValueSome struct (Span<Code>([||]), CbAst.Union (AUnion("asd", CbAst.Union (AUnion ("123", CbAst.Bool true))))), r)

File: TestParser/TestUnion.cs
using CbStyles.Cbon;
using CbStyles.Cbon.Parser;
using NUnit.Framework;

namespace TestParser
{
    public class TestUnion
    {
        [SetUp]
        public void Setup() { }

        [Test]
        public void TestUnion1()
        {
            var code = "(asd) 123";
            Parser.RunInReader(code, code => {
                var r = Parser.Union(code).Value.Item2;
                Assert.AreEqual(new CbUnion("asd", CbVal.NewNum("123")), r);
            });
        }

        [Test]
        public void TestUnion2()
        {
            var code = "('asd')(123)true";
            Parser.RunInReader(code, code => {
                var r = Parser.Union(code).Value.Item2;
                Assert.AreEqual(new CbUnion("asd", CbVal.NewUnion("123", CbVal.NewBool(true))), r);
            });
        }

        [Test]
        public void TestUnion3()
        {
            var code = "()";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.Union(code);
            }));
            Assert.AreEqual("Expected union tag but not found \t at 1:1", e.Message);
        }

        [Test]
        public void TestUnion4()
        {
            var code = "(asd";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.Union(code);
            }));
            Assert.AreEqual("Tag not close \t at 1:4", e.Message);
        }

        [Test]
        public void TestUnion5()
        {
            var code = "(asd)";
            var e = Assert.Throws<ParserException>(() => Parser.RunInReader(code, code => {
                Parser.Union(code);
            }));
            Assert.AreEqual("Unexpected EOF \t at 1:5", e.Message);
        }
    }
}
